src/Voron/Impl/PagerState.cs

```diff
@@ -100,24 +100,75 @@ public PagerState(AbstractPager pager, long prefetchSegmentSize, long prefetchRe
 
         public readonly byte* MapBase; // { get; set; }
 
-        internal bool _released;
-
         public void Release()
         {
-            if (Interlocked.Decrement(ref _refs) != 0)
+            int currentRefCount;
+            int newRefCount;
+
+            // Uses atomic compare-and-swap to safely decrement the reference count,
+            // ensuring consistency even with concurrent modifications by other threads.
+            do
+            {
+                currentRefCount = _refs;
+
+                // Checks for negative reference counts to detect potential misuse or logical errors,
+                // such as double releases or illegal state transitions, preventing further incorrect decrements.
+                if (currentRefCount < 0)
+                    ThrowInvalidPagerState();
+
+                newRefCount = currentRefCount - 1;
+            }
+            while (Interlocked.CompareExchange(ref _refs, newRefCount, currentRefCount) != currentRefCount);
+
+            // If the reference count is still positive, there are active references.
+            if (newRefCount > 0)
```

Comment:
This is problematic, you may end up with _two_ threads entering this code, then calling `ReleaseInternal`

Reply:
Took me the most part of 4 hours to reason about this particular code a more than a handful of rewrites. 

The key is to realize that can only be true if there bugs in the code. The reason why that is so is because of observability rules at `#line 124`. Since we are capturing the value of `_refs` at `#line 112' and then check if the state is invalid (either there is a bug OR it is already disposed) then at `#line 124` we would observe a sequence of values where the following property holds true for the entire sequence: `v(n) != v(n+1)`

Because of that, we do know that errors in usage of this reference counter would push us into negative territory and cause `#line 117` to trigger inside the loop (`#line 129` is there just in case that logic is wrong and that's why only runs in debug). 

Unless I am missing something there is no way we can observe at `#line 124` two times the value 0 because at the moment that `_refs` becomes `0` the `PagerState` is already as good as dead.

Reply:
It is valid for you to get `currentRefCount == 0`, in which case you'll decrement to -1 and then in release call `ReleaseInternal` twice.

If it is already `0` when you called `Release` is a problem, for sure, but the previous code handled that, I would rather not lose it.

We need to protect ourselves from accidental double free, etc.